dist/types/types.d.ts

```diff
@@ -0,0 +1,78 @@
+import { ParsedQs } from "qs";
+/**
+ * Offers the route lookup to be merged with declarations emitted from ts generator
+ */
+export interface RouteLookup {
+}
+declare type _ParameterBindingDefinition = {
+    name: string;
+    binding: string;
+} | {
+    name: string;
+};
+declare type _BindingOrValue<T extends _ParameterBindingDefinition> = T extends {
+    binding: string;
+} ? ParameterBinding<T['binding']> : ParameterValue;
+/** Type of the values a parameter can be passed as. */
+export declare type ParameterValue = string | number;
+/** A parameter as object or primitive */
+export declare type ParameterBinding<Name extends string> = {
+    [K in Name]: ParameterValue;
+} | ParameterValue;
+/** Ordered array containing the paramters required to hydrate a route */
+export declare type RouteParamsArray<Name extends RouteName> = Name extends keyof RouteLookup ? _RouteParamsArray<RouteLookup[Name]> : GenericRouteParamsArray;
+declare type _RouteParamsArray<T extends readonly _ParameterBindingDefinition[]> = {
+    [K in keyof T]: _BindingOrValue<T[K]> | null;
+};
+/** Object containing the paramters required to hydrate a route.  */
+export declare type RouteParamsObject<Name extends RouteName> = Name extends keyof RouteLookup ? _RouteParamsObject<RouteLookup[Name]> & QueryObject : GenericRouteParamsObject & QueryObject;
+declare type _RouteParamsObject<T extends readonly _ParameterBindingDefinition[]> = {
+    [Item in T[number] as Item['name']]?: _BindingOrValue<Item> | null;
+};
+/** Object containing the normalized paramters required to hydrate a route.  */
+export declare type NormalizedRouteParams<Route extends RouteName> = Route extends keyof RouteLookup ? Record<keyof RouteParamsObject<Route>, ParameterValue> & _NormalizedRouteParams : _NormalizedRouteParams;
+export declare type _NormalizedRouteParams = Record<keyof any, ParameterValue> & QueryObject;
+export declare type GenericRouteParamsArray = _BindingOrValue<_ParameterBindingDefinition>[];
+export declare type GenericRouteParamsObject = QueryObject & Record<keyof any, any>;
+export declare type GenericRouteParams = GenericRouteParamsArray | GenericRouteParamsObject | ParameterValue;
+export declare type QueryObject = {
+    _query?: Record<string, any>;
+};
+/**
+ * Holds any known route name or an arbitrary string.
+ */
+export declare type RouteName = keyof RouteLookup | (string & {});
```

Comment:
@lmeysel what does `& {}` do here? If we want to be permissive and allow strings we don't recognize as route names, why not just `keyof RouteLookup | string`?

Reply:
Hmm, not sure if this question is answered in the other PR already: You will need `string & {}` to make (at least VSCode) to allow strings as well as "specific" strings (i.e. literals). If you just do `| string`, vscode will make no suggestions, and thus you will not get a list of available routes in intellisense